### PR TITLE
Optional sorting of predicates

### LINKED ISSUE #206: Make query sort optional
It was discovered in Exawind that sorting is way too expensive for those meshes. There is no easy way to disable it. Right now, I have a hacky patch:
```diff
--- a/src/details/ArborX_DetailsBoundingVolumeHierarchyImpl.hpp
+++ b/src/details/ArborX_DetailsBoundingVolumeHierarchyImpl.hpp
@@ -384,9 +384,15 @@ BoundingVolumeHierarchyImpl<DeviceType>::queryDispatch(
 
   Kokkos::Profiling::pushRegion("ArborX:BVH:sort_queries");
 
+#if 0
   auto const permute =
       Details::BatchedQueries<DeviceType>::sortQueriesAlongZOrderCurve(
           bvh.bounds(), predicates);
+#else
+  Kokkos::View<size_t *> permute(
+      Kokkos::ViewAllocateWithoutInitializing("permute"), n_queries);
+  iota(permute);
+#endif
 
   // FIXME  readability!  queries is a sorted copy of the predicates
   auto queries = Details::BatchedQueries<DeviceType>::applyPermutation(
```
We need a proper API for this.

### PR DESCRIPTION
**Pull request by aprokop**
Fix #206.

I added a new boolean parameter. I opted to have it before `buffer_size` as it is common from both kNN and spatial searches, while the  parameters after it are specific (`buffer_size` to spatial, `which` to kNN). This choice made the patch slightly larger than it could have been.

I also decided to keep the default to sorted as of now. I'm not married to it, but this allows us to still be able to compare benchmark results with older runs.

### CODE REVIEW

**Comment by aprokop** [comment]
Just to note: running with sorting disabled indeed makes our standard benchmark slower. But we knew that.

File: test/tstLinearBVH.cpp [hunk 1]
```diff
@@ -740,9 +806,8 @@ BOOST_AUTO_TEST_CASE_TEMPLATE(structured_grid, DeviceType, ARBORX_DEVICE_TYPES)
 
   ArborX::BVH<DeviceType> bvh(bounding_boxes);
 
-  // (i) use same objects for the queries than the objects we constructed the
-  // BVH
-  // i-2  i-1  i  i+1
+  // (i) use same objects for the queries than the objects we constructed
+  // the BVH i-2  i-1  i  i+1
```
Comment (dalg24) [inline test/tstLinearBVH.cpp]: Undo this change

Comment (aprokop) [reply test/tstLinearBVH.cpp]: I did not do it manually, it was `clang-format`.

**Comment by aprokop** [comment]
> Did you consider having a struct that agglomerate all possible options?

I did. I don't see much difference, and then one to also have to remember which parameters make sense in which function. I think longer term this indeed would be the right way to go, though. If you feel strongly about it, I could do that (introduce `TraversalPolicy` or something similar).

File: test/tstLinearBVH.cpp (hunk 1 above)
Comment (dalg24) [reply test/tstLinearBVH.cpp]: Well you tell clang-format I don't like it and revert :p

Comment (aprokop) [reply test/tstLinearBVH.cpp]: OK

**Comment by aprokop** [comment]
This is ready for review. @dalg24 I addressed your comments.

File: src/details/ArborX_DetailsBoundingVolumeHierarchyImpl.hpp [hunk 2]
```diff
@@ -72,6 +72,35 @@ enum class NearestQueryAlgorithm
...
 };
 
+} // namespace Details
+
+struct TraversalPolicy
+{
+  int _buffer_size = 0;
+  bool _do_predicate_sort = true;
+  Details::NearestQueryAlgorithm _which =
+      Details::NearestQueryAlgorithm::StackBased_Default;
```
Comment (dalg24) [inline src/details/ArborX_DetailsBoundingVolumeHierarchyImpl.hpp]: Name is a bit silly for a data member (not very descriptive)

File: src/details/ArborX_DetailsBoundingVolumeHierarchyImpl.hpp [hunk 3]
```diff
@@ -72,6 +72,35 @@ enum class NearestQueryAlgorithm
   PriorityQueueBased_Deprecated
 };
 
+} // namespace Details
+
+struct TraversalPolicy
+{
+  int _buffer_size = 0;
+  bool _do_predicate_sort = true;
```
Comment (dalg24) [inline src/details/ArborX_DetailsBoundingVolumeHierarchyImpl.hpp]: I'd prefer `sort_predicates`

File: src/details/ArborX_DetailsBoundingVolumeHierarchyImpl.hpp [hunk 4]
```diff
@@ -72,6 +72,35 @@ enum class NearestQueryAlgorithm
   PriorityQueueBased_Deprecated
 };
 
+} // namespace Details
+
+struct TraversalPolicy
```
[the remark this answers is not in the document]
Comment (aprokop) [reply src/details/ArborX_DetailsBoundingVolumeHierarchyImpl.hpp]: I disagree with `Details::` as we expect users to use it eventially. `Experimental` sounds reasonable.

File: src/details/ArborX_DetailsBoundingVolumeHierarchyImpl.hpp [hunk 5]
```diff
@@ -72,6 +72,35 @@ enum class NearestQueryAlgorithm
...
+  bool _do_predicate_sort = true;
+  Details::NearestQueryAlgorithm _which =
+      Details::NearestQueryAlgorithm::StackBased_Default;
+
+  TraversalPolicy &setBufferSize(int buffer_size)
+  {
+    _buffer_size = buffer_size;
+    return *this;
+  }
+  TraversalPolicy &setTraversalAlgorithm(Details::NearestQueryAlgorithm which)
```
[the remark this answers is not in the document]
Comment (aprokop) [reply src/details/ArborX_DetailsBoundingVolumeHierarchyImpl.hpp]: I'm thinking that this could be used later to set traversal algorithms for both spatial and nearest approaches, so don't want to narrow it down.

File: src/details/ArborX_DetailsBoundingVolumeHierarchyImpl.hpp (hunk 2 above)
Comment (aprokop) [reply src/details/ArborX_DetailsBoundingVolumeHierarchyImpl.hpp]: That's fair, I was not thinking, just copy-pasting the original `which` parameter.

File: src/details/ArborX_DetailsBoundingVolumeHierarchyImpl.hpp (hunk 3 above)
Comment (aprokop) [reply src/details/ArborX_DetailsBoundingVolumeHierarchyImpl.hpp]: OK.
